minari/utils/data_collector.py

```diff
@@ -0,0 +1,573 @@
+            )
+        # create local directory if it doesn't exist
+        if not os.path.exists(self.datasets_path):
+            os.makedirs(self.datasets_path)
+
+        self._tmp_dir = tempfile.TemporaryDirectory(dir=self.datasets_path)
+        self._tmp_f = h5py.File(
+            os.path.join(self._tmp_dir.name, "tmp_dataset.hdf5"), "a", track_order=True
+        )  # track insertion order of groups ('episodes_i')
+        self._tmp_f.attrs[
+            "env_spec"
+        ] = self.env.spec.to_json()  # pyright: ignore [reportOptionalMemberAccess]
+        self._tmp_f.attrs[
+            "flatten_observation"
+        ] = self._step_data_callback.flatten_observation
+        self._tmp_f.attrs["flatten_action"] = self._step_data_callback.flatten_action
+
+        self._new_episode = False
+        self._eps_group = None
+
+        # Initialize first episode group in temporary hdf5 file
+        self._episode_id = 0
+        self._eps_group: h5py.Group = self._tmp_f.create_group("episode_0")
+        self._eps_group.attrs["id"] = 0
+
+        self._last_episode_group_term_or_trunc = False
+        self._last_episode_n_steps = 0
+
+    def _add_to_episode_buffer(
+        self, episode_buffer: Dict[str, Union[list, Dict]], step_data: Dict
+    ) -> Dict:
+        """Add step data dictionary to episode buffer.
+
+        Args:
+            buffer (Dict): dictionary episode buffer
+            step_data (Dict): dictionary with data for a single step
+
+        Returns:
+            Dict: new dictionary episode buffer with added values from step_data
+        """
+        for key, value in step_data.items():
+            if not self._record_infos and key == "infos":
+                continue
+            if key not in episode_buffer:
+                if isinstance(value, dict):
+                    episode_buffer[key] = self._add_to_episode_buffer({}, value)
+                else:
+                    episode_buffer[key] = [value]
+            else:
+                if isinstance(value, dict):
+                    episode_buffer[key] = self._add_to_episode_buffer(
+                        episode_buffer[key], value
+                    )
+                else:
+                    episode_buffer[key].append(value)
+
+        return episode_buffer
+
+    def step(self, action: Any):
+        """Gymnasium step method."""
+        obs, rew, terminated, truncated, info = self.env.step(action)
+
+        # add/edit data from step and convert to dictionary step data
+        step_data = self._step_data_callback(
+            env=self,
+            obs=obs,
+            info=info,
+            action=action,
+            rew=rew,
+            terminated=terminated,
+            truncated=truncated,
+        )
+
+        # force step data dicitonary to include keys corresponding to Gymnasium step returns:
+        # actions, observations, rewards, terminations, truncatins, and infos
+        assert STEP_DATA_KEYS.issubset(step_data.keys())
```

Comment:
`assert isinstance(step_data, StepData)`

Reply:
I think `issubset` works better in this case since `step_data` can have extra keys apart from the `STEP_DATA_KEYS`. `STEP_DATA_KEYS` are mandatory keys that have to be returned every step.